include/ccf/indexing/seqnos_by_key.h

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/indexing/strategy.h"
+#include "indexing/indexing_types.h"
+
+#include <memory>
+#include <string>
+
+namespace ccf::indexing::strategies
+{
+  template <typename M>
+  class SeqnosByKey : public Strategy
+  {
+  private:
+    // Key is the raw value of a KV key.
+    // Value is every SeqNo which talks about that key.
+    // This entire structure should be shardable, so we can dump parts of it to
+    // disk!
+    std::unordered_map<kv::untyped::SerialisedEntry, SeqNoCollection>
+      seqnos_by_key;
+
+    ccf::TxID current_txid = {};
+
+    std::string map_name;
+
+  public:
+    SeqnosByKey(const std::string& map_name_) :
+      Strategy(fmt::format("SeqnosByKey for {}", map_name_)),
+      map_name(map_name_)
+    {}
+
+    SeqnosByKey(const M& map) : SeqnosByKey(map.get_name()) {}
+
+    void handle_committed_transaction(
+      const ccf::TxID& tx_id, const StorePtr& store) override
+    {
+      // NB: Don't use M, instead get an untyped view over the map with the same
+      // name. This saves deserialisation here, where we work with the raw key.
+      auto tx = store->create_tx();
+      auto handle = tx.ro<kv::untyped::Map>(map_name);
+      handle->foreach(
+        [this, seqno = tx_id.seqno](const auto& k, const auto& v) {
+          seqnos_by_key[k].insert(seqno);
+          return true;
+        });
+      current_txid = tx_id;
+    }
+
+    ccf::TxID get_indexed_watermark() const
+    {
+      return current_txid;
+    }
+
+    SeqNoCollection get_all_write_txs(const typename M::Key& key)
+    {
+      const auto serialised_key = M::KeySerialiser::to_serialised(key);
+      const auto it = seqnos_by_key.find(serialised_key);
+      if (it != seqnos_by_key.end())
+      {
+        return it->second;
+      }
+
+      return {};
```

Comment:
```suggestion
      return = SeqNoCollection();
```

For consistency with line :106

Reply:
Sure, I've changed this for consistency. The instance on line :106 is actually required where this one is auto-derived, because there we're deliberately returning a _non-empty `optional`, containing an empty `SeqNoCollection`_ (where `return {}` would just be an empty `optional`).